File: Week1/linkedListCycle/linkedListCycle.js

//Definition for singly - linked list.
function ListNode(val) {
    this.val = val;
    this.next = null;
}


/**
 * @param {ListNode} head
 * @return {boolean}
 */

var hasCycle = function (head) {
    const hasSeen = new Set();
    let currentNode = head;
    while (currentNode !== null) {
        if (hasSeen.has(currentNode)) {
            return true
        } else {
            hasSeen.add(currentNode)
        }
        currentNode = currentNode.next
    }
    return false;
};

File: Week1/linkedListCycle/linkedListCycle.py
from types import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        hasSeen = {}
        currentNode = head
        while currentNode:
            if currentNode in hasSeen:
                return True
            else:
                hasSeen[currentNode] = True
            currentNode = currentNode.next
        return False


#the turtle and hair solution: O(1) space solution
class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        slow, fast = head, head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                return True
            
        return False